File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.js
/**
 * @param {string[]} words
 * @return {number}
 */
const reverse = str => str.split("").reverse().join("");

const longestPalindrome = (words) => {
  const obj = {};
  let counter = 0;

  for (const word of words) {
    const reversedWord = reverse(word);

    if (obj[word]) {
      obj[word]--;
      counter += 4;
    } else {
      obj[reversedWord] = obj[reversedWord] + 1 || 1;
    }
  }

  const hashGotMorePalindromes = Object.keys(obj).filter(str => obj[str] && reverse(str) === str);

  if (hashGotMorePalindromes.length) counter += 2;

  return counter;
};

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.cpp
class Solution
{
public:
  int longestPalindrome(vector<string> &words)
  {
    unordered_map<string, int> map;
    int counter = 0;

    for (string w : words)
    {
      string rev = w;
      reverse(w.begin(), w.end());
      if (map[w])
      {
        map[w]--;
        counter += 4;
      }
      else
      {
        map[rev] += 1;
      }
    }
    vector<string> moreWords{};
    for (auto &&i : map)
    {
      string first = i.first;
      string rev = first;
      reverse(first.begin(), first.end());
      if (map[first] && rev == first)
      {
        moreWords.push_back(first);
      }
    }
    if (moreWords.size() > 0)
    {
      counter += 2;
    }

    return counter;
  }
};